Automattic/woocommerce-payments
# Add feature flag check to early WooPay session request

#### Changes proposed in this Pull Request


This PR is a follow up to #7455. In that PR we disable WooPay's first party authentication feature flag (`_wcpay_feature_woopay_first_party_auth`) when incompatible plugins are enabled. This prevented the first party flow from firing successfully and used the OTP modal instead. Unfortunately, it didn't fix the primary issue which was to allow WooPay compatible extensions to pass their data on to WooPay due to the changes in #6831. 

This PR fixes that issue by wrapping the changes from #6831 in a check against the feature flag as well, so that that feature flag will determine whether we send any early session data to WooPay or not. The key piece to avoid when working with compatible extensions (adapted extensions) is that we must have the shopper's email address before we initialize the session, something we can only do in most cases after the shopper has authenticated on WooPay.







#### Testing instructions

For testing these changes specifically:

- Monitor the wp_options table for the option name _wcpay_feature_woopay_first_party_auth.
- Remove the option if it is present.
- Disable any adapted extensions (tested with [Gift Cards](https://woocommerce.com/products/gift-cards/) and [Points and Rewards](https://woocommerce.com/products/woocommerce-points-and-rewards/)).
- Confirm that the option is now set to 1.
- Enable either/all of the extensions listed.
- Confirm that the option is now set to 0.
- Add a product to the Cart
- Click any of the WooPay express checkout buttons
- Confirm that there was no call to `?wc-ajax=wcpay_get_woopay_session`
- Confirm that the OTP modal popped up instead of redirecting to WooPay right away

- Disable all the adapted extensions.
- Confirm that the option is now set to 1.
- Add a product to the Cart
- Click any of the WooPay express checkout buttons
- Confirm that there was a call to `?wc-ajax=wcpay_get_woopay_session`
- Confirm that you were redirected directly to WooPay and there was no WooPay modal.

Testing the plugins still in compatiblity.

- Repeat the steps above until you get to having both plugins enabled and the option set to 0.
- Then follow these testing instructions for Gift Cards: https://github.com/Automattic/woopay/pull/1954
- Then follow these testing instructions for Points and Rewards: https://github.com/Automattic/woopay/pull/1950
- With these tests you should be able to complete a full checkout in both cases.
-------------------


**Post merge**
